tutorial/write-custom-widget.md

```diff
@@ -107,3 +107,184 @@ Take care not to keep any important state in your renderer - animation tickers
+	return newMyWidgetRenderer(w.minSize, w.text)
+}
+```
+
+The widget is now almost complete except for the ```newMyWidgetRenderer()``` function which creates your new widget renderer. 
+
+**Warning**: Do not keep a reference to the widget renderer in the widget (or any other part of your application). The fyne application caches the renderer and may destroy it at any time. The fyne application will call ```CreateRenderer()``` again if it requires a new renderer instance. It is ok for the widget renderer to refer to the widget but not the other way round.
+
+#### Create a Widget Renderer
+
+The Widget Renderer is how we tell the fyne application what to draw and how to lay it out.
+
+The Widget Renderer has it's own state and methods (see ```fyne.WidgetRenderer``` interface) but it should only be concerned with drawing (rendering) the widget.
+
+The Widget Renderer is responsible for:
+
+1) Defining the objects that are to be drawn. These are returned by the ```Objects() []fyne.CanvasObject``` method. Each object returned will need to be created by the Widget Renderer and initialised before being returned.
+3) The position and size of all canvas items within the space passed in to the Layout method.
+3) Ensuring that the display is refreshed with changes to the state of the widget. 
+
+Our example renderer could be stored in the following struct:
+
+```
+type myWidgetRenderer struct {
+	widget     *MyWidget
+	background *canvas.Rectangle
+	text       *canvas.Text
+}
+```
+
+As you can see it is all about canvas objects that the fyne application will render.
+
+For this example we need a canvas.Rectangle as a background and a canvas.Text to display the widget text.
+
+We hold a reference to the widget so we can refer to the data it holds. If the data in the widget changes we want the renderer to be able to refresh the display. See the Refresh() method below.
+
+When we construct the renderer we pass in a reference to the widget and store it.
+
+```go
+func newMyWidgetRenderer(myWidget *MyWidget) *myWidgetRenderer {
+	return &myWidgetRenderer{
+		widget:     myWidget,
+		background: canvas.NewRectangle(theme.BackgroundColor()),
+		text:       canvas.NewText(myWidget.text, theme.ForegroundColor()),
+	}
+}
+```
+
+Notice that the size and position of the canvas objects are not defined here, only initial colours and text properties.
+
+The Objects() method should return the canvas objects in a list as follows:
+
+```go
+func (r *myWidgetRenderer) Objects() []fyne.CanvasObject {
+    // The order is critical, rect is drawn first then text
+	return []fyne.CanvasObject{r.background, r.text}
+}
+```
+
+Each object returned will need to be sized and positioned in the Layout() method as follows:
+
+```go
+func (r *myWidgetRenderer) Layout(s fyne.Size) {
+	ts := fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
+	r.text.Move(fyne.Position{X: (s.Width - ts.Width) / 2, Y: (s.Height - ts.Height) / 2})
+	r.background.Resize(s)
+}
+```
+
```

Comment:
I wonder if this bulleted list would benefit from a short sentence leading in?
Maybe "As you can see in the code sample, our `Layout` method arranges are objects by doing:".